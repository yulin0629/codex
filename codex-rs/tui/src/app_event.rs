use codex_core::protocol::Event;
use crossterm::event::KeyEvent;

use crate::slash_command::SlashCommand;

#[allow(clippy::large_enum_variant)]
pub(crate) enum AppEvent {
    CodexEvent(Event),

    Redraw,

    KeyEvent(KeyEvent),

    /// Scroll event with a value representing the "scroll delta" as the net
    /// scroll up/down events within a short time window.
    Scroll(i32),

    /// Request to exit the application gracefully.
    ExitRequest,

    /// Forward an `Op` to the Agent. Using an `AppEvent` for this avoids
    /// bubbling channels through layers of widgets.
    CodexOp(codex_core::protocol::Op),

    /// Latest formatted log line emitted by `tracing`.
    LatestLog(String),

    /// Dispatch a recognized slash command from the UI (composer) to the app
    /// layer so it can be handled centrally.
    DispatchCommand(SlashCommand),
}
