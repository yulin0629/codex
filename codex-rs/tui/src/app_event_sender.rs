use std::sync::mpsc::Sender;

use crate::app_event::AppEvent;

#[derive(Clone, Debug)]
pub(crate) struct AppEventSender {
    app_event_tx: Sender<AppEvent>,
}

impl AppEventSender {
    pub(crate) fn new(app_event_tx: Sender<AppEvent>) -> Self {
        Self { app_event_tx }
    }

    /// Send an event to the app event channel. If it fails, we swallow the
    /// error and log it.
    pub(crate) fn send(&self, event: AppEvent) {
        if let Err(e) = self.app_event_tx.send(event) {
            tracing::error!("failed to send event: {e}");
        }
    }
}
