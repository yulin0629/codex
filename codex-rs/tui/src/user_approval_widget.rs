//! A modal widget that prompts the user to approve or deny an action
//! requested by the agent.
//!
//! This is a (very) rough port of
//! `src/components/chat/terminal-chat-command-review.tsx` from the TypeScript
//! UI to Rust using [`ratatui`]. The goal is feature‑parity for the keyboard
//! driven workflow – a fully‑fledged visual match is not required.

use std::path::PathBuf;

use codex_core::protocol::Op;
use codex_core::protocol::ReviewDecision;
use crossterm::event::KeyCode;
use crossterm::event::KeyEvent;
use ratatui::buffer::Buffer;
use ratatui::layout::Rect;
use ratatui::prelude::*;
use ratatui::text::Line;
use ratatui::text::Span;
use ratatui::widgets::Block;
use ratatui::widgets::BorderType;
use ratatui::widgets::Borders;
use ratatui::widgets::List;
use ratatui::widgets::Paragraph;
use ratatui::widgets::Widget;
use ratatui::widgets::WidgetRef;
use tui_input::Input;
use tui_input::backend::crossterm::EventHandler;

use crate::app_event::AppEvent;
use crate::app_event_sender::AppEventSender;
use crate::exec_command::relativize_to_home;
use crate::exec_command::strip_bash_lc_and_escape;

/// Request coming from the agent that needs user approval.
pub(crate) enum ApprovalRequest {
    Exec {
        id: String,
        command: Vec<String>,
        cwd: PathBuf,
        reason: Option<String>,
    },
    ApplyPatch {
        id: String,
        reason: Option<String>,
        grant_root: Option<PathBuf>,
    },
}

/// Options displayed in the *select* mode.
struct SelectOption {
    label: &'static str,
    decision: Option<ReviewDecision>,
    /// `true` when this option switches the widget to *input* mode.
    enters_input_mode: bool,
}

// keep in same order as in the TS implementation
const SELECT_OPTIONS: &[SelectOption] = &[
    SelectOption {
        label: "Yes (y)",
        decision: Some(ReviewDecision::Approved),

        enters_input_mode: false,
    },
    SelectOption {
        label: "Yes, always approve this exact command for this session (a)",
        decision: Some(ReviewDecision::ApprovedForSession),

        enters_input_mode: false,
    },
    SelectOption {
        label: "Edit or give feedback (e)",
        decision: None,

        enters_input_mode: true,
    },
    SelectOption {
        label: "No, and keep going (n)",
        decision: Some(ReviewDecision::Denied),

        enters_input_mode: false,
    },
    SelectOption {
        label: "No, and stop for now (esc)",
        decision: Some(ReviewDecision::Abort),

        enters_input_mode: false,
    },
];

/// Internal mode the widget is in – mirrors the TypeScript component.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum Mode {
    Select,
    Input,
}

/// A modal prompting the user to approve or deny the pending request.
pub(crate) struct UserApprovalWidget<'a> {
    approval_request: ApprovalRequest,
    app_event_tx: AppEventSender,
    confirmation_prompt: Paragraph<'a>,

    /// Currently selected index in *select* mode.
    selected_option: usize,

    /// State for the optional input widget.
    input: Input,

    /// Current mode.
    mode: Mode,

    /// Set to `true` once a decision has been sent – the parent view can then
    /// remove this widget from its queue.
    done: bool,
}

// Number of lines automatically added by ratatui’s [`Block`] when
// borders are enabled (one at the top, one at the bottom).
const BORDER_LINES: u16 = 2;

impl UserApprovalWidget<'_> {
    pub(crate) fn new(approval_request: ApprovalRequest, app_event_tx: AppEventSender) -> Self {
        let input = Input::default();
        let confirmation_prompt = match &approval_request {
            ApprovalRequest::Exec {
                command,
                cwd,
                reason,
                ..
            } => {
                let cmd = strip_bash_lc_and_escape(command);
                // Maybe try to relativize to the cwd of this process first?
                // Will make cwd_str shorter in the common case.
                let cwd_str = match relativize_to_home(cwd) {
                    Some(rel) => format!("~/{}", rel.display()),
                    None => cwd.display().to_string(),
                };
                let mut contents: Vec<Line> = vec![
                    Line::from("Shell Command".bold()),
                    Line::from(""),
                    Line::from(vec![
                        format!("{cwd_str}$").dim(),
                        Span::from(format!(" {cmd}")),
                    ]),
                    Line::from(""),
                ];
                if let Some(reason) = reason {
                    contents.push(Line::from(reason.clone().italic()));
                    contents.push(Line::from(""));
                }
                contents.extend(vec![Line::from("Allow command?"), Line::from("")]);
                Paragraph::new(contents)
            }
            ApprovalRequest::ApplyPatch {
                reason, grant_root, ..
            } => {
                let mut contents: Vec<Line> =
                    vec![Line::from("Apply patch".bold()), Line::from("")];

                if let Some(r) = reason {
                    contents.push(Line::from(r.clone().italic()));
                    contents.push(Line::from(""));
                }

                if let Some(root) = grant_root {
                    contents.push(Line::from(format!(
                        "This will grant write access to {} for the remainder of this session.",
                        root.display()
                    )));
                    contents.push(Line::from(""));
                }

                contents.push(Line::from("Allow changes?"));
                contents.push(Line::from(""));

                Paragraph::new(contents)
            }
        };

        Self {
            approval_request,
            app_event_tx,
            confirmation_prompt,
            selected_option: 0,
            input,
            mode: Mode::Select,
            done: false,
        }
    }

    pub(crate) fn get_height(&self, area: &Rect) -> u16 {
        let confirmation_prompt_height =
            self.get_confirmation_prompt_height(area.width - BORDER_LINES);

        match self.mode {
            Mode::Select => {
                let num_option_lines = SELECT_OPTIONS.len() as u16;
                confirmation_prompt_height + num_option_lines + BORDER_LINES
            }
            Mode::Input => {
                //   1. "Give the model feedback ..." prompt
                //   2. A single‑line input field (we allocate exactly one row;
                //      the `tui-input` widget will scroll horizontally if the
                //      text exceeds the width).
                const INPUT_PROMPT_LINES: u16 = 1;
                const INPUT_FIELD_LINES: u16 = 1;

                confirmation_prompt_height + INPUT_PROMPT_LINES + INPUT_FIELD_LINES + BORDER_LINES
            }
        }
    }

    fn get_confirmation_prompt_height(&self, width: u16) -> u16 {
        // Should cache this for last value of width.
        self.confirmation_prompt.line_count(width) as u16
    }

    /// Process a `KeyEvent` coming from crossterm. Always consumes the event
    /// while the modal is visible.
    /// Process a key event originating from crossterm. As the modal fully
    /// captures input while visible, we don’t need to report whether the event
    /// was consumed—callers can assume it always is.
    pub(crate) fn handle_key_event(&mut self, key: KeyEvent) {
        match self.mode {
            Mode::Select => self.handle_select_key(key),
            Mode::Input => self.handle_input_key(key),
        }
    }

    fn handle_select_key(&mut self, key_event: KeyEvent) {
        match key_event.code {
            KeyCode::Up => {
                if self.selected_option == 0 {
                    self.selected_option = SELECT_OPTIONS.len() - 1;
                } else {
                    self.selected_option -= 1;
                }
            }
            KeyCode::Down => {
                self.selected_option = (self.selected_option + 1) % SELECT_OPTIONS.len();
            }
            KeyCode::Char('y') => {
                self.send_decision(ReviewDecision::Approved);
            }
            KeyCode::Char('a') => {
                self.send_decision(ReviewDecision::ApprovedForSession);
            }
            KeyCode::Char('n') => {
                self.send_decision(ReviewDecision::Denied);
            }
            KeyCode::Char('e') => {
                self.mode = Mode::Input;
            }
            KeyCode::Enter => {
                let opt = &SELECT_OPTIONS[self.selected_option];
                if opt.enters_input_mode {
                    self.mode = Mode::Input;
                } else if let Some(decision) = opt.decision {
                    self.send_decision(decision);
                }
            }
            KeyCode::Esc => {
                self.send_decision(ReviewDecision::Abort);
            }
            _ => {}
        }
    }

    fn handle_input_key(&mut self, key_event: KeyEvent) {
        // Handle special keys first.
        match key_event.code {
            KeyCode::Enter => {
                let feedback = self.input.value().to_string();
                self.send_decision_with_feedback(ReviewDecision::Denied, feedback);
            }
            KeyCode::Esc => {
                // Cancel input – treat as deny without feedback.
                self.send_decision(ReviewDecision::Denied);
            }
            _ => {
                // Feed into input widget for normal editing.
                let ct_event = crossterm::event::Event::Key(key_event);
                self.input.handle_event(&ct_event);
            }
        }
    }

    fn send_decision(&mut self, decision: ReviewDecision) {
        self.send_decision_with_feedback(decision, String::new())
    }

    fn send_decision_with_feedback(&mut self, decision: ReviewDecision, _feedback: String) {
        let op = match &self.approval_request {
            ApprovalRequest::Exec { id, .. } => Op::ExecApproval {
                id: id.clone(),
                decision,
            },
            ApprovalRequest::ApplyPatch { id, .. } => Op::PatchApproval {
                id: id.clone(),
                decision,
            },
        };

        // Ignore feedback for now – the current `Op` variants do not carry it.

        // Forward the Op to the agent. The caller (ChatWidget) will trigger a
        // redraw after it processes the resulting state change, so we avoid
        // issuing an extra Redraw here to prevent a transient frame where the
        // modal is still visible.
        self.app_event_tx.send(AppEvent::CodexOp(op));
        self.done = true;
    }

    /// Returns `true` once the user has made a decision and the widget no
    /// longer needs to be displayed.
    pub(crate) fn is_complete(&self) -> bool {
        self.done
    }
}

const PLAIN: Style = Style::new();
const BLUE_FG: Style = Style::new().fg(Color::Blue);

impl WidgetRef for &UserApprovalWidget<'_> {
    fn render_ref(&self, area: Rect, buf: &mut Buffer) {
        // Take the area, wrap it in a block with a border, and divide up the
        // remaining area into two chunks: one for the confirmation prompt and
        // one for the response.
        let outer = Block::default()
            .title("Review")
            .borders(Borders::ALL)
            .border_type(BorderType::Rounded);
        let inner = outer.inner(area);
        let prompt_height = self.get_confirmation_prompt_height(inner.width);
        let chunks = Layout::default()
            .direction(Direction::Vertical)
            .constraints([Constraint::Length(prompt_height), Constraint::Min(0)])
            .split(inner);
        let prompt_chunk = chunks[0];
        let response_chunk = chunks[1];

        // Build the inner lines based on the mode. Collect them into a List of
        // non-wrapping lines rather than a Paragraph because get_height(Rect)
        // depends on this behavior for its calculation.
        let lines = match self.mode {
            Mode::Select => SELECT_OPTIONS
                .iter()
                .enumerate()
                .map(|(idx, opt)| {
                    let (prefix, style) = if idx == self.selected_option {
                        ("▶", BLUE_FG)
                    } else {
                        (" ", PLAIN)
                    };
                    Line::styled(format!("  {prefix} {}", opt.label), style)
                })
                .collect(),
            Mode::Input => {
                vec![
                    Line::from("Give the model feedback on this command:"),
                    Line::from(self.input.value()),
                ]
            }
        };

        outer.render(area, buf);
        self.confirmation_prompt.clone().render(prompt_chunk, buf);
        Widget::render(List::new(lines), response_chunk, buf);
    }
}
