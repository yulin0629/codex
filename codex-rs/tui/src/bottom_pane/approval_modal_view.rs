use crossterm::event::KeyEvent;
use ratatui::buffer::Buffer;
use ratatui::layout::Rect;
use ratatui::widgets::WidgetRef;

use crate::app_event_sender::AppEventSender;
use crate::user_approval_widget::ApprovalRequest;
use crate::user_approval_widget::UserApprovalWidget;

use super::BottomPane;
use super::BottomPaneView;

/// Modal overlay asking the user to approve/deny a sequence of requests.
pub(crate) struct ApprovalModalView<'a> {
    current: UserApprovalWidget<'a>,
    queue: Vec<ApprovalRequest>,
    app_event_tx: AppEventSender,
}

impl ApprovalModalView<'_> {
    pub fn new(request: ApprovalRequest, app_event_tx: AppEventSender) -> Self {
        Self {
            current: UserApprovalWidget::new(request, app_event_tx.clone()),
            queue: Vec::new(),
            app_event_tx,
        }
    }

    pub fn enqueue_request(&mut self, req: ApprovalRequest) {
        self.queue.push(req);
    }

    /// Advance to next request if the current one is finished.
    fn maybe_advance(&mut self) {
        if self.current.is_complete() {
            if let Some(req) = self.queue.pop() {
                self.current = UserApprovalWidget::new(req, self.app_event_tx.clone());
            }
        }
    }
}

impl<'a> BottomPaneView<'a> for ApprovalModalView<'a> {
    fn handle_key_event(&mut self, _pane: &mut BottomPane<'a>, key_event: KeyEvent) {
        self.current.handle_key_event(key_event);
        self.maybe_advance();
    }

    fn is_complete(&self) -> bool {
        self.current.is_complete() && self.queue.is_empty()
    }

    fn calculate_required_height(&self, area: &Rect) -> u16 {
        self.current.get_height(area)
    }

    fn render(&self, area: Rect, buf: &mut Buffer) {
        (&self.current).render_ref(area, buf);
    }

    fn try_consume_approval_request(&mut self, req: ApprovalRequest) -> Option<ApprovalRequest> {
        self.enqueue_request(req);
        None
    }
}
