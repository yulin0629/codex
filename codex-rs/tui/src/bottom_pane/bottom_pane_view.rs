use crate::user_approval_widget::ApprovalRequest;
use crossterm::event::KeyEvent;
use ratatui::buffer::Buffer;
use ratatui::layout::Rect;

use super::BottomPane;

/// Type to use for a method that may require a redraw of the UI.
pub(crate) enum ConditionalUpdate {
    NeedsRedraw,
    NoRedraw,
}

/// Trait implemented by every view that can be shown in the bottom pane.
pub(crate) trait BottomPaneView<'a> {
    /// Handle a key event while the view is active. A redraw is always
    /// scheduled after this call.
    fn handle_key_event(&mut self, _pane: &mut BottomPane<'a>, _key_event: KeyEvent) {}

    /// Return `true` if the view has finished and should be removed.
    fn is_complete(&self) -> bool {
        false
    }

    /// Height required to render the view.
    fn calculate_required_height(&self, area: &Rect) -> u16;

    /// Render the view: this will be displayed in place of the composer.
    fn render(&self, area: Rect, buf: &mut Buffer);

    /// Update the status indicator text.
    fn update_status_text(&mut self, _text: String) -> ConditionalUpdate {
        ConditionalUpdate::NoRedraw
    }

    /// Called when task completes to check if the view should be hidden.
    fn should_hide_when_task_is_done(&mut self) -> bool {
        false
    }

    /// Try to handle approval request; return the original value if not
    /// consumed.
    fn try_consume_approval_request(
        &mut self,
        request: ApprovalRequest,
    ) -> Option<ApprovalRequest> {
        Some(request)
    }
}
