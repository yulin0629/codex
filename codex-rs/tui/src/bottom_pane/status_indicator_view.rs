use ratatui::buffer::Buffer;
use ratatui::layout::Rect;
use ratatui::widgets::WidgetRef;

use crate::app_event_sender::AppEventSender;
use crate::status_indicator_widget::StatusIndicatorWidget;

use super::BottomPaneView;
use super::bottom_pane_view::ConditionalUpdate;

pub(crate) struct StatusIndicatorView {
    view: StatusIndicatorWidget,
}

impl StatusIndicatorView {
    pub fn new(app_event_tx: AppEventSender, height: u16) -> Self {
        Self {
            view: StatusIndicatorWidget::new(app_event_tx, height),
        }
    }

    pub fn update_text(&mut self, text: String) {
        self.view.update_text(text);
    }
}

impl<'a> BottomPaneView<'a> for StatusIndicatorView {
    fn update_status_text(&mut self, text: String) -> ConditionalUpdate {
        self.update_text(text);
        ConditionalUpdate::NeedsRedraw
    }

    fn should_hide_when_task_is_done(&mut self) -> bool {
        true
    }

    fn calculate_required_height(&self, _area: &Rect) -> u16 {
        self.view.get_height()
    }

    fn render(&self, area: Rect, buf: &mut Buffer) {
        self.view.render_ref(area, buf);
    }
}
