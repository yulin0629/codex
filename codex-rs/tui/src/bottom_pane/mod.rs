//! Bottom pane: shows the ChatComposer or a BottomPaneView, if one is active.

use bottom_pane_view::BottomPaneView;
use bottom_pane_view::ConditionalUpdate;
use crossterm::event::KeyEvent;
use ratatui::buffer::Buffer;
use ratatui::layout::Rect;
use ratatui::widgets::WidgetRef;

use crate::app_event::AppEvent;
use crate::app_event_sender::AppEventSender;
use crate::user_approval_widget::ApprovalRequest;

mod approval_modal_view;
mod bottom_pane_view;
mod chat_composer;
mod chat_composer_history;
mod command_popup;
mod status_indicator_view;

pub(crate) use chat_composer::ChatComposer;
pub(crate) use chat_composer::InputResult;

use approval_modal_view::ApprovalModalView;
use status_indicator_view::StatusIndicatorView;

/// Pane displayed in the lower half of the chat UI.
pub(crate) struct BottomPane<'a> {
    /// Composer is retained even when a BottomPaneView is displayed so the
    /// input state is retained when the view is closed.
    composer: ChatComposer<'a>,

    /// If present, this is displayed instead of the `composer`.
    active_view: Option<Box<dyn BottomPaneView<'a> + 'a>>,

    app_event_tx: AppEventSender,
    has_input_focus: bool,
    is_task_running: bool,
}

pub(crate) struct BottomPaneParams {
    pub(crate) app_event_tx: AppEventSender,
    pub(crate) has_input_focus: bool,
}

impl BottomPane<'_> {
    pub fn new(params: BottomPaneParams) -> Self {
        Self {
            composer: ChatComposer::new(params.has_input_focus, params.app_event_tx.clone()),
            active_view: None,
            app_event_tx: params.app_event_tx,
            has_input_focus: params.has_input_focus,
            is_task_running: false,
        }
    }

    /// Forward a key event to the active view or the composer.
    pub fn handle_key_event(&mut self, key_event: KeyEvent) -> InputResult {
        if let Some(mut view) = self.active_view.take() {
            view.handle_key_event(self, key_event);
            if !view.is_complete() {
                self.active_view = Some(view);
            } else if self.is_task_running {
                let height = self.composer.calculate_required_height(&Rect::default());
                self.active_view = Some(Box::new(StatusIndicatorView::new(
                    self.app_event_tx.clone(),
                    height,
                )));
            }
            self.request_redraw();
            InputResult::None
        } else {
            let (input_result, needs_redraw) = self.composer.handle_key_event(key_event);
            if needs_redraw {
                self.request_redraw();
            }
            input_result
        }
    }

    /// Update the status indicator text (only when the `StatusIndicatorView` is
    /// active).
    pub(crate) fn update_status_text(&mut self, text: String) {
        if let Some(view) = &mut self.active_view {
            match view.update_status_text(text) {
                ConditionalUpdate::NeedsRedraw => {
                    self.request_redraw();
                }
                ConditionalUpdate::NoRedraw => {
                    // No redraw needed.
                }
            }
        }
    }

    /// Update the UI to reflect whether this `BottomPane` has input focus.
    pub(crate) fn set_input_focus(&mut self, has_focus: bool) {
        self.has_input_focus = has_focus;
        self.composer.set_input_focus(has_focus);
    }

    pub fn set_task_running(&mut self, running: bool) {
        self.is_task_running = running;

        match (running, self.active_view.is_some()) {
            (true, false) => {
                // Show status indicator overlay.
                let height = self.composer.calculate_required_height(&Rect::default());
                self.active_view = Some(Box::new(StatusIndicatorView::new(
                    self.app_event_tx.clone(),
                    height,
                )));
                self.request_redraw();
            }
            (false, true) => {
                if let Some(mut view) = self.active_view.take() {
                    if view.should_hide_when_task_is_done() {
                        // Leave self.active_view as None.
                        self.request_redraw();
                    } else {
                        // Preserve the view.
                        self.active_view = Some(view);
                    }
                }
            }
            _ => {
                // No change.
            }
        }
    }

    /// Called when the agent requests user approval.
    pub fn push_approval_request(&mut self, request: ApprovalRequest) {
        let request = if let Some(view) = self.active_view.as_mut() {
            match view.try_consume_approval_request(request) {
                Some(request) => request,
                None => {
                    self.request_redraw();
                    return;
                }
            }
        } else {
            request
        };

        // Otherwise create a new approval modal overlay.
        let modal = ApprovalModalView::new(request, self.app_event_tx.clone());
        self.active_view = Some(Box::new(modal));
        self.request_redraw()
    }

    /// Height (terminal rows) required by the current bottom pane.
    pub fn calculate_required_height(&self, area: &Rect) -> u16 {
        if let Some(view) = &self.active_view {
            view.calculate_required_height(area)
        } else {
            self.composer.calculate_required_height(area)
        }
    }

    pub(crate) fn request_redraw(&self) {
        self.app_event_tx.send(AppEvent::Redraw)
    }

    /// Returns true when the slash-command popup inside the composer is visible.
    pub(crate) fn is_command_popup_visible(&self) -> bool {
        self.active_view.is_none() && self.composer.is_command_popup_visible()
    }

    // --- History helpers ---

    pub(crate) fn set_history_metadata(&mut self, log_id: u64, entry_count: usize) {
        self.composer.set_history_metadata(log_id, entry_count);
    }

    pub(crate) fn on_history_entry_response(
        &mut self,
        log_id: u64,
        offset: usize,
        entry: Option<String>,
    ) {
        let updated = self
            .composer
            .on_history_entry_response(log_id, offset, entry);

        if updated {
            self.request_redraw();
        }
    }
}

impl WidgetRef for &BottomPane<'_> {
    fn render_ref(&self, area: Rect, buf: &mut Buffer) {
        // Show BottomPaneView if present.
        if let Some(ov) = &self.active_view {
            ov.render(area, buf);
        } else {
            (&self.composer).render_ref(area, buf);
        }
    }
}
