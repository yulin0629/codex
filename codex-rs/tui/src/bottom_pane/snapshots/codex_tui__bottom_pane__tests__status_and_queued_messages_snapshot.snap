---
source: tui/src/bottom_pane/mod.rs
expression: "render_snapshot(&pane, area)"
---
                                                
• Working (0s • esc to interrupt)               
  ↳ Queued follow-up question                   
    ⌥ + ↑ edit                                  
                                                
                                                
› Ask Codex to do anything                      
                                                
  100% context left · ? for shortcuts
